abstract_domains/usage/used.py

```diff
@@ -0,0 +1,115 @@
+from abstract_domains.lattice import Lattice
+from enum import Flag
+
+
+class Used(Flag):
+    """Used state of a program variable."""
+    # do not change values blindly, they are used for easy implementation with bitwise operators
+    U = 3
+    UU = 2
+    UN = 1
+    N = 0
+
+
+U = Used.U
+UU = Used.UU
+UN = Used.UN
+N = Used.N
+
+
+class UsedLattice(Lattice):
+    """Used variable analysis core abstract domain representation."""
+
+    _DESCEND = {
+        U: UU,
+        UU: UU,
+        UN: N,
+        N: N
+    }
+
+    _COMBINE = {
+        (N, N): N,
+        (N, UU): UU,
+        (N, UN): UN,
+        (N, U): U,
+
+        (UU, N): UU,
+        (UU, UU): UU,
+        (UU, UN): UN,
+        (UU, U): U,
+
+        (UN, N): UN,
+        (UN, UU): UU,
+        (UN, UN): UN,
+        (UN, U): U,
+
+        (U, N): U,
+        (U, UU): U,
+        (U, UN): UN,
+        (U, U): U,
+    }
```

Comment:
What kind of code construct are these _DESCEND and _COMBINE? I have never seen such a thing :D